### PR TITLE
ROX-24978: Compatibility tests should not test eol combinations

### PR DESCRIPTION
**Pull request by jschnath**
### Description

We now use an API call to figure out which versions are supported instead of testing against the last n in our gke-version-compatibility-tests

### User-facing documentation
- [x] CHANGELOG update is not needed
- [x] Documentation is not needed

### Testing

- [x] inspected CI results

#### Automated testing

- [x] added compatibility tests

#### How I validated my change

I manually ran the new python function and confirmed that it returned the correct versions (only 4.4 right now)

Compatibility tests are now testing nothing, this is intended. Refer to the product lifecycles API here:
https://access.redhat.com/product-life-cycles/api/v1/products?name=Red%20Hat%20Advanced%20Cluster%20Security%20for%20Kubernetes
Maintenance support is a new category introduced in 4.4 so once we release 4.5 OR add a support exception compatibility tests will have versions to test against again.

### CODE REVIEW

**Comment by vikin91** [comment]
> Compatibility tests are now testing nothing, this is intended

I think, we should still test compatibility of the current master against the latest 4.4.

File: scripts/ci/jobs/gke_version_compatibility_tests.py
Comment (vikin91) [inline scripts/ci/jobs/gke_version_compatibility_tests.py:63]: Has this been confirmed with Marina? The expiration is only 5 days old, maybe there are some talks to extend it further?

**Comment by jschnath** [comment]
> I think, we should still test compatibility of the current master against the latest 4.4.

I just checked some artifacts of previous compatibility runs and it looks like we were doing that before, so I'll change the code to do this again, eg. always test the current build against the latest release.

File: scripts/ci/jobs/gke_version_compatibility_tests.py
Comment (jschnath) [reply scripts/ci/jobs/gke_version_compatibility_tests.py:63]: I'll ask her 👍

Comment (jschnath) [reply scripts/ci/jobs/gke_version_compatibility_tests.py:63]: She's on vacation but will be back to work on Monday according to her Slack status.

**Comment by jschnath** [comment]
image
@vikin91 I've re-added the support exception for now and will probably remove it in another PR once I have confirmation from Marina that it was not extended. We now also test latest release vs current commit.

**Comment by jschnath** [comment]
Just got word from Marina that the support exception is indeed close and won't be extended. Removing the exception again.

**Comment by jschnath** [comment]
Now testing only latest release vs current commit, see [Artifacts](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/stackrox_stackrox/11791/pull-ci-stackrox-stackrox-master-gke-version-compatibility-tests/1810299554243809280/artifacts/gke-version-compatibility-tests/stackrox-stackrox-e2e-test/artifacts/)


image

**Comment by jschnath** [comment]
Compatibility tests have been broken for 4 days now, eg here's the job history in the [nightlies](https://prow.ci.openshift.org/job-history/gs/test-platform-results/logs/branch-ci-stackrox-stackrox-nightlies-gke-version-compatibility-tests).

However the exact state of code currently in this PR has previously passed all tests, so I don't think those failures should stand in the way of reviewing / merging this.

**Comment by jschnath** [comment]
Currently blocked until https://github.com/stackrox/stackrox/pull/11946 is resolved.

Upstream kernel driver build have been removed so we can no longer use GKE driver builds for 3.74, but Giles is now looking into reinstating 3.74 without collector which should hopefully work. Will rebase once that one is merged.

**Comment by jschnath** [comment]
No longer blocked, rebasing

**Comment by jschnath** [comment]
@vikin91 unblocked, rebased, and all green. Good for another review when you have the time.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:136]: I would check that the keys and the length of the arrays are valid before attempting to access the values here.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:150]: I would also make sure this success before returning.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:72]: Note: I can't comment outside of the changed lines.

[nit] I know this was not changed in this PR but I saw a `print` in line 64. I think we should changed to `logging`.

File: scripts/ci/jobs/gke_version_compatibility_tests.py
Comment (lvalerom) [inline scripts/ci/jobs/gke_version_compatibility_tests.py:45]: [nit] I kind of like having the `ifs` separated like they were. That way if for whatever reason there's a change in the way we retrieve the charts and we land in a situation where we have charts for central but not for sensor (or the other way around) we can log exactly what is going on.

**Review by lvalerom — COMMENTED** [review]
I think it looks good just some minor comments. I'll give it a green after I test it (I haven't done that yet 😅 )

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: Done

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: Sure, I'll add some exception handling.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py:72]: Okay, changing it. It's not that important since it's only called within main() which is used to debug the script / check if it's working correctly but it's better nonetheless.

File: scripts/ci/jobs/gke_version_compatibility_tests.py
Comment (jschnath) [reply scripts/ci/jobs/gke_version_compatibility_tests.py:45]: The reason I changed this is that with the product lifecycles API support for versions between central and sensor is now always symmetrical. 

> That way if for whatever reason there's a change in the way we retrieve the charts and we land in a situation where we have charts for central but not for sensor (or the other way around) we can log exactly what is going on.

This should already be logged in `__get_supported_helm_chart_versions` where it prints a `debug.logging` message for each central or sensor chart that is supported but not available in helm

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:139]: If you want to go the python way and use the `try-catch`, I would normally catch the specific exceptions this piece of code can throw; in this case `KeyError` and `IndexError`.

You can also check if a particular key is in a dictionary with `"key" in data`. Similarly with the length of the array you can use `len(array) > 0`.With this you can avoid the `try-catch`.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:141]: This one is not handled.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:162]: Same here. I think `urlopen` raises `URLError`

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:159]: `JSONDecodeError`. Not sure if can raise other exceptions though

File: scripts/ci/jobs/gke_version_compatibility_tests.py
Comment (lvalerom) [reply scripts/ci/jobs/gke_version_compatibility_tests.py:45]: > is now always symmetrical

My point exactly, `now` is symmetrical but it might not be at some point. I think my initial point was confusing because I wrote `the way we retrieve the charts` when I meant `the way we retrieve the supported versions`.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (lvalerom) [inline .openshift-ci/get_latest_helm_chart_versions.py:153]:
```suggestion
        with urlopen(req) as response:
```

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: This is one thing I dislike about python, there is no good way to know which exceptions a library function may throw without reading all the code. I've looked at the documentation for the functions I'm using and am trying to narrow it down now, but of course this is no _guarantee_ for safety.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: Done

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: Yes it does, I just checked the documentation

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: I don't think it can from what I just googled 👍

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: I do have to split urlopen(req) out of the with if I want to put it into a try / except block. But I have restructured the code so it makes more sense.

Comment (lvalerom) [reply .openshift-ci/get_latest_helm_chart_versions.py]: 🤔 I tried with the code I suggested below I think it works inside the `try-catch`. I'm not sure if the `with` is used properly the way you have here. As far as I recall `with` is used with classes that have `__init__`, `__enter__`, `__exit__` defined or with functions that have a `yield` function. I think this scenario is the later so you need the `with urlopen(req) as request:` in order for the `with` to work properly.

Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: I've read up a bit about `with` in python, you're right. I'll put `urlopen(req)` back inside it and the try except block around it.

File: .openshift-ci/get_latest_helm_chart_versions.py
Comment (jschnath) [reply .openshift-ci/get_latest_helm_chart_versions.py]: Done

**Review by lvalerom — COMMENTED** [review]
LGTM. Thanks for the improvements. I think we should run the compatibility tests again just in case before merging though

**Comment by jschnath** [comment]
compatibility tests passing and testing the expected versions @lvalerom I think this can be merged

image

**Review by lvalerom — APPROVED** [review]
Awesome! Thanks again for the improvements 🙂

File: scripts/ci/jobs/gke_version_compatibility_tests.py
Comment (porridge) [inline scripts/ci/jobs/gke_version_compatibility_tests.py:78]: @jschnath this change seems to have diverged the comment from the code. The comment now claims there are no support exceptions, but the code immediately after it does add a support exception item. What am I missing?

Comment (jschnath) [reply scripts/ci/jobs/gke_version_compatibility_tests.py:78]: There was a support exception as linked in https://issues.redhat.com/browse/ROX-18223 that expired, but a new one / an extension to support for this version was being negotiated at the time I made this commit if I recall correctly. And this would have been before my 1 year long parental leave.

I just did a search for it now and it seems the support exception is still active (as of June 12th I found it mentioned in slack), see https://redhat-internal.slack.com/archives/C033SKHG9CZ/p1781284756201909?thread_ts=1781214312.906519&cid=C033SKHG9CZ

**Pull request merged**
